Implement support for crypto/ed25519 using the OpenSSL backend (#1066)
microsoft/go: eng/doc/fips/UserGuide.md

### [crypto/ed25519](https://pkg.go.dev/crypto/ed25519)

Package ed25519 implements the Ed25519 signature algorithm. See https://ed25519.cr.yp.to/.

**Requirements**

The CNG backend and some old OpenSSL distributions don't support ED25519.
In those cases, the code will fall back to standard Go crypto.

#### func [GenerateKey](https://pkg.go.dev/crypto/ed25519#GenerateKey)

```go
func GenerateKey(rand io.Reader) (pub ed25519.PublicKey, priv ed25519.PrivateKey, error)
```

GenerateKey generates a public/private key pair using entropy from rand. If rand is nil, crypto/rand.Reader will be used.

**Requirements**

- `rand` must be boring.RandReader or nil, else GenerateKey will panic. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

> `pub` and `priv` are generated using using [EVP_PKEY_keygen] using the `EVP_PKEY_ED25519` algorithm.

Reviewer: `using using [EVP_PKEY_keygen] using` has a duplicate but I also think it might be a bit easier on the eyes to avoid sandwiching `[EVP_PKEY_keygen]` with the same word twice:

```suggestion
`pub` and `priv` are generated using [EVP_PKEY_keygen] with the `EVP_PKEY_ED25519` algorithm.
```

</details>

#### func [Sign](https://pkg.go.dev/crypto/ed25519#Sign)

```go
func Sign(privateKey ed25519.PrivateKey, message []byte) []byte
```

Sign signs the message with privateKey and returns a signature. It will panic if len(privateKey) is not PrivateKeySize.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

> `message` is signed using [EVP_MD_CTX_new], [EVP_DigestSignInit] and [EVP_DigestSign].

Reviewer: `EVP_DigestSignInit` missing a URL below.

</details>

#### func [Verify](https://pkg.go.dev/crypto/ed25519#Verify)

```go
func Verify(publicKey ed25519.PublicKey, message, sig []byte) bool
```

Verify reports whether sig is a valid signature of message by publicKey. It will panic if len(publicKey) is not PublicKeySize.

**Requirements**

> - Falls back to standard Go code when using OpenSSL 1.1.1a or lower.

Reviewer: We established earlier that requirement = FIPS compatibility, so I think we can make this more focused on that goal and flip the condition. (IMO this makes the doc more consistent/easier to read.)

```suggestion
- OpenSSL version must be 1.1.1b or higher. Otherwise, falls back to standard Go crypto.
```

In theory, we can leave off the fallback description too because the intro says that we fall back unless otherwise specified, but I could go either way on that. 😄

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

> `message` is verified against `sig` using [EVP_MD_CTX_new], [DigestVerifyInit] and [EVP_DigestVerify].

Reviewer: ```suggestion
`message` is verified against `sig` using [EVP_MD_CTX_new], [EVP_DigestVerifyInit] and [EVP_DigestVerify].

```

</details>

#### func [VerifyWithOptions](https://pkg.go.dev/crypto/ed25519#VerifyWithOptions)

```go
func VerifyWithOptions(publicKey PublicKey, message, sig []byte, opts *Options) error
```

VerifyWithOptions reports whether sig is a valid signature of message by publicKey. A valid signature is indicated by returning a nil error. It will panic if len(publicKey) is not PublicKeySize.

**Requirements**

- Only `opts.Hash == nil && opts.Context == ""` is implemented using the OpenSSL backend. Other combinations fall back to standard Go code.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

> `message` is verified against `sig` using [EVP_MD_CTX_new], [DigestVerifyInit] and [EVP_DigestVerify].

Reviewer: ```suggestion
`message` is verified against `sig` using [EVP_MD_CTX_new], [EVP_DigestVerifyInit] and [EVP_DigestVerify].

```

</details>

#### func [NewKeyFromSeed](https://pkg.go.dev/crypto/ed25519#NewKeyFromSeed)

```go
func NewKeyFromSeed(seed []byte) (priv ed25519.PrivateKey)
```

NewKeyFromSeed calculates a private key from a seed. It will panic if len(seed) is not SeedSize.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

> `priv` is generated using using [EVP_PKEY_new_raw_private_key] using the `EVP_PKEY_ED25519` algorithm.

Reviewer: ```suggestion
`priv` is generated using [EVP_PKEY_new_raw_private_key] with the `EVP_PKEY_ED25519` algorithm.

```

[EVP_DigestFinal]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestFinal.html
[EVP_DigestInit]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestInit.html
[EVP_DigestInit_ex]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestInit_ex.html
[EVP_DigestSign]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestSign.html
[EVP_DigestVerify]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestVerify.html

> [EVP_DigestSign]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestSign.html

Reviewer: Missing link above:

```suggestion
[EVP_DigestSign]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestSign.html
[EVP_DigestSignInit]: https://www.openssl.org/docs/man3.0/man3/EVP_DigestSignInit.html
```

(It looks like the OpenSSL docs are down at the moment, everything's a 404 even navigating from their front page or google results.)
